qiskit/transpiler/coupling.py

```diff
@@ -463,20 +463,38 @@ def connected_components(self) -> List["CouplingMap"]:
 
         will print ``3`` as index ``0`` in the second component is qubit 3 in the original cmap.
 
+        Args:
+            target (Target): If used the provided :class:`~.Target` will be used to filter any
+                qubits that do not have any supported operations so that those qubits
+                will not be included as a connected component in the output list. When not
+                specified these qubits without any supported operations will be included in
+                the output as a connected component of a single qubit.
+
         Returns:
             list: A list of :class:`~.CouplingMap` objects for each connected
                 components. The order of this list is deterministic but
                 implementation specific and shouldn't be relied upon as
                 part of the API.
         """
+        graph = self.graph
+        if target is not None:
+            qargs = target.qargs
+            graph = self.graph.copy()
+            for index in graph.node_indices():
+                for qargs in target.qargs:
+                    if index in qargs:
+                        break
+                else:
+                    graph.remove_node(index)
```

Comment:
This seems a bit confusing (we have both `qargs` and direct calls to `target.qargs`) and inefficient - we scan through all qargs in the target for each of the node indices in the happy path (no faulty gates), which feels a bit quadratic.  Does something like
```python
has_operations = set(itertools.chain.from_iterable(target.qargs))
for index in graph.node_indices():
    if index not in has_operations:
       graph.remove_node(index)
```

I'm not 100% certain that that's the same logic, though.

Reply:
I'm confused about the purpose of  `qargs = target.qargs`in line 481. The loop variable `qargs` below is not introduced in a new scope, so it overwrites the previous definition in case `graph.node_indices()` is not empty. And otherwise, it is never used.

Reply:
I reworked the logic using your suggestion as part of doing the filtering in the target as part of: https://github.com/Qiskit/qiskit-terra/commit/c6a1e89d5211f9cd7cbcc70320272014a22cfcc7